sw/device/lib/testing/spi_flash_emulator.c

```diff
@@ -0,0 +1,160 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/testing/spi_flash_emulator.h"
+
+#include <stdint.h>
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/spi_device_testutils.h"
+#include "sw/device/lib/testing/spi_flash_testutils.h"
+
+enum {
+  // JEDEC standard continuation code.
+  kJedecContCode = 0x7f,
+  // JEDEC ID of lowRISC.
+  kJedecManufacturer = 0xEF,
+  // lowRISC is on page 12 of JEDEC IDs.
+  kJedecContCodeCount = 12,
+  // A density of 24 corresponds to 16MiB (1<<24 bytes).
+  kJedecDensity = 24,
+  // The standard SFDP signature value.
+  kSfdpSignature = 0x50444653,
+};
+
+typedef struct sfdp_header {
+  uint32_t signature;
+  uint8_t minor;
+  uint8_t major;
+  uint8_t nph;
+  uint8_t reserved;
+} sfdp_header_t;
+
+typedef struct parameter_header {
+  uint8_t param_id;
+  uint8_t minor;
+  uint8_t major;
+  uint8_t length;
+  uint8_t table_pointer[3];
+  uint8_t pad;
+} parameter_header_t;
+
+typedef struct bfpt {
+  uint32_t data[9];
+} bfpt_t;
+
+typedef struct sfdp {
+  sfdp_header_t header;
+  parameter_header_t param;
+  bfpt_t bfpt;
+} sfdp_t;
+
+static status_t prepare_sfdp(dif_spi_host_t *spih,
+                             dif_spi_device_handle_t *spid) {
+  uint8_t data[256];
+  spi_flash_testutils_read_sfdp(spih, 0, data, sizeof(data));
+
+  // TODO: present a better SFDP table.  This table is as simple
+  // as possible and copies only a few bits of relevant data
+  // from the backend EEPROM's table.
+  sfdp_t sfdp = {
+      .header = {.signature = kSfdpSignature,
+                 .minor = 0,
+                 .major = 1,
+                 .nph = 0,
+                 .reserved = 0xFF},
+      .param =
+          {
+              .param_id = 0x00,
+              .minor = 0,
+              .major = 1,
+              .length = sizeof(bfpt_t) / sizeof(uint32_t),
+              .table_pointer = {(uint8_t)offsetof(sfdp_t, bfpt)},
+              .pad = 0xFF,
+          },
+      .bfpt = {0},
+  };
+
+  uint32_t offset =
+      read_32(data + offsetof(sfdp_t, param.table_pointer)) & 0x00FFFFFF;
```

Comment:
nit: I suspect `data` doesn't have a larger alignment requirement than 1 byte--Typically, aggregates have an alignment requirement that is the same as their unit type's.

If we want to use `read_32`, it'd be best to make that happen, either via a `uint32_t` array or an `alignas` specification. Probably the `alignas` is the simplest for avoiding any pointer arithmetic confusion.

Reply:
Fixed.